pkg/upstream/proxy.go

```diff
@@ -55,19 +69,107 @@ func (m *multiUpstreamProxy) ServeHTTP(rw http.ResponseWriter, req *http.Request
-	m.serveMux.Handle(upstream.Path, newHTTPUpstreamProxy(upstream, u, sigData, writer.ProxyErrorHandler))
+	return m.registerHandler(upstream, newHTTPUpstreamProxy(upstream, u, sigData, writer.ProxyErrorHandler), writer)
+}
+
+// registerHandler ensures the given handler is regiestered with the serveMux.
+func (m *multiUpstreamProxy) registerHandler(upstream options.Upstream, handler http.Handler, writer pagewriter.Writer) error {
+	if upstream.RewriteTarget == "" {
+		m.registerSimpleHandler(upstream.Path, handler)
+		return nil
+	}
+
+	return m.registerRewriteHandler(upstream, handler, writer)
+}
+
+// registerSimpleHandler maintains the behaviour of the go standard serveMux
+// by ensuring any path with a trailing `/` matches all paths under that prefix.
+func (m *multiUpstreamProxy) registerSimpleHandler(path string, handler http.Handler) {
+	if strings.HasSuffix(path, "/") {
+		m.serveMux.PathPrefix(path).Handler(handler)
+	} else {
+		m.serveMux.Path(path).Handler(handler)
+	}
+}
+
+// registerRewriteHandler ensures the handler is registered for all paths
+// which match the regex defined in the Path.
+// Requests to the handler will have the request path rewritten before the
+// request is made to the next handler.
+func (m *multiUpstreamProxy) registerRewriteHandler(upstream options.Upstream, handler http.Handler, writer pagewriter.Writer) error {
+	rewriteRegExp, err := regexp.Compile(upstream.Path)
+	if err != nil {
+		return fmt.Errorf("invalid path %q for upstream: %v", upstream.Path, err)
+	}
+
+	rewrite := newRewritePath(rewriteRegExp, upstream.RewriteTarget, writer)
+	h := alice.New(rewrite).Then(handler)
+	m.serveMux.MatcherFunc(func(req *http.Request, match *mux.RouteMatch) bool {
+		return rewriteRegExp.MatchString(req.URL.Path)
+	}).Handler(h)
+
+	return nil
+}
+
+// registerTrailingSlashHandler creates a new matcher that will check if the
+// requested path would match if it had a trailing slash appended.
+// If the path matches with a trailing slash, we send back a redirect.
+// This allows us to be consistent with the built in go servemux implementation.
+func registerTrailingSlashHandler(serveMux *mux.Router) {
+	serveMux.MatcherFunc(func(req *http.Request, _ *mux.RouteMatch) bool {
+		if strings.HasSuffix(req.URL.Path, "/") {
+			return false
+		}
+
+		// Use a separate RouteMatch so that we can redirect to the path + /.
+		// If we pass through the match then the matched backed will be served
+		// instead of the redirect handler.
+		m := &mux.RouteMatch{}
+		slashReq := req.Clone(context.Background())
+		slashReq.URL.Path += "/"
+		return serveMux.Match(slashReq, m)
+	}).Handler(http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		http.Redirect(rw, req, req.URL.String()+"/", http.StatusMovedPermanently)
+	}))
+}
+
+// sortByPathLongest ensures that the upstreams are sorted by longest path.
+// If rewrites are involved, a rewrite takes precedence over a non-rewrite.
```

Comment:
Can you elaborate on this aspect? What happens if the rewrite makes it shorter?

Reply:
We don't care about what happens if the rewrite makes it shorter, we just register it based on the longest regex (ie so it should probably match fewer things). The idea is shorter paths and shorter regexes match more entries and so they need to be evaluated later